src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,120 @@
+            /// </summary>
+            /// <remarks>
+            /// An empty action set is still allowed and will return a modified solution
+            /// that will update the document properties as appropriate.
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new InvalidOperationException("Cannot apply action that is not in ApplicableActions");
+                }
+
+                // Prior to updating the solution it's possible the document id has changed between the time 
+                // the document action info was generated and when the solution update is applied. We
+                // do a best effort to still locate the document if the id has changed.
+                var document = GetDocument(solution);
+
+                // If the document was found in the solution then the current id will be durable across actions
+                // since we own the solution snapshot at this point. 
+                var documentId = document.Id;
+
+                // Make sure that the document name and folders are updated to what we expect them to be
+                solution = solution
+                    .WithDocumentName(documentId, _documentName)
+                    .WithDocumentFolders(documentId, _documentFolders);
+
+                // Apply each action individually. Order should not matter
+                foreach (var action in actions)
+                {
+                    document = solution.GetRequiredDocument(documentId);
+                    solution = await action.GetModifiedSolutionAsync(document, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            /// <summary>
+            /// Attempts to find the document in the solution. Tries by documentId first, but 
+            /// that's not always reliable between analysis and application of the rename actions
+            /// </summary>
+            private Document GetDocument(Solution solution)
+            {
+                // DocumentId is the best bet for finding a document,
+                // but it's possible the document was renamed or moved
+                // before actions were applied.
+                if (solution.ContainsDocument(_documentId))
+                {
+                    return solution.GetRequiredDocument(_documentId);
+                }
+
+                var project = solution.GetRequiredProject(_projectId);
+                return project.Documents.FirstOrDefault(d => d.Name == _documentName && d.Folders.SequenceEqual(_documentFolders))
+                    ?? throw new InvalidOperationException("Unable to find document that was used for analysis in current solution");
```

Comment:
this seems strange.  why do we need this fallback at all?

Reply:
Since the actions themselves have to work on a solution that comes in after analysis, it's possible (and in some cases expected) that the document id is not the same.

The use case that brought the need for this was how the project system is expected to consume these APIs

1. Call `RenameDocumentAsync`
2. Call current APIs to remove document and add new document to tracking
3. Display UI to confirm user wants to apply actions (if any exist)
4. Apply actions

Since 2. above is the current way that documents are renamed, we expect the document id to be regenerated. If we fix that issue, we wouldn't have a known use for this fallback.